File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/22-PQ/LeetCode/LeetCode-2558.java
class Solution {
    /// O(klogn + nlogn) algorithm with O(n) extra space
    public long pickGifts(int[] gifts, int k) {
        PriorityQueue<Integer> pq = new PriorityQueue<>((x, y)->y-x);
        for (var gift: gifts) pq.offer(gift);

        while (k-- > 0){
            var num = pq.poll();
            pq.offer((int)Math.sqrt(num));
        } // end-while

        long result = 0;
        for (var num: pq) result += num;
        return result;
    }
}

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/22-PQ/LeetCode/LeetCode-2558.cpp
class Solution {
public:
    /// O(klogn + nlogn) algorithm with O(n) extra space
    long long pickGifts(vector<int>& gifts, int k) {
        std::priority_queue<int> Q(gifts.begin(), gifts.end());
        while (k--){
            auto num = Q.top();
            Q.pop();
            Q.push((int)sqrt(num));
        } // end-while

        int64_t result = 0;
        while (Q.size()){
            auto num = Q.top();
            Q.pop();
            result += num;
        } // end-while

        return result;
    } // end-pickGifts
};